core/commands/mfs/mfs.go

```diff
@@ -0,0 +1,689 @@
+		cmds.StringArg("name", true, false, "name of filesystem to close"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		node, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		if node.Mfs == nil {
+			res.SetError(cmds.ErrNotOnline, cmds.ErrNormal)
+			return
+		}
+
+		name := req.Arguments()[0]
+
+		final, err := node.Mfs.CloseRoot(name)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		res.SetOutput(&Object{Hash: final.B58String()})
+	},
+	Marshalers: cmds.MarshalerMap{
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+			out := res.Output().(*Object)
+			return strings.NewReader(out.Hash), nil
+		},
+	},
+	Type: Object{},
+}
+
+type MfsLsOutput struct {
+	Entries []mfs.NodeListing
+}
+
+var MfsLsCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline:          "List directories inside a filesystem",
+		ShortDescription: ``,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("path", true, false, "path to show listing for"),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("l", "use long listing format"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		dir, err := getRootDir(req)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		path := req.Arguments()[0]
+		if len(path) > 0 && path[0] == '/' {
+			path = path[1:]
+		}
+
+		var nd mfs.FSNode
+		if len(path) > 0 {
+			foundnd, err := mfs.DirLookup(dir, path)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			nd = foundnd
+		} else {
+			nd = dir
+		}
+
+		switch nd := nd.(type) {
+		case *mfs.Directory:
+			listing, err := nd.List()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			res.SetOutput(&MfsLsOutput{listing})
+			return
+		case *mfs.File:
+			parts := strings.Split(path, "/")
+			name := parts[len(parts)-1]
+			out := &MfsLsOutput{[]mfs.NodeListing{mfs.NodeListing{Name: name, Type: 1}}}
+			res.SetOutput(out)
+			return
+		default:
+			res.SetError(errors.New("unrecognized type"), cmds.ErrNormal)
+		}
+	},
+	Marshalers: cmds.MarshalerMap{
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+			out := res.Output().(*MfsLsOutput)
+			buf := new(bytes.Buffer)
+			long, _, _ := res.Request().Option("l").Bool()
+
+			for _, o := range out.Entries {
+				if long {
+					fmt.Fprintf(buf, "%s\t%s\t%d\n", o.Name, o.Hash, o.Size)
+				} else {
+					fmt.Fprintf(buf, "%s\n", o.Name)
+				}
+			}
+			return buf, nil
+		},
+	},
+	Type: MfsLsOutput{},
+}
+
+var MfsPutCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "import a given file into a filesystem",
+		ShortDescription: `
+Mfs 'put' takes a file that already exists in ipfs and imports it into a
+given mfs filesystem.
+		`,
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("object", true, false, "ipfspath of object to import"),
+		cmds.StringArg("path", true, false, "path within filesystem to import to"),
```

Comment:
maybe can detect whether it's an ipfs path with `/ipfs/` as the prefix? (get users thinking about `/ipfs/...` and their local fs coexisting)